src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs

```diff
@@ -0,0 +1,458 @@
+
+                    case "tokens":
+                    case "requests":
+
+                        var retryAfter = TimeSpan.Zero;
+                        if (headers.Contains("retry-after-ms") == false)
+                        {
+                            throw new GenAiTooManyTokensException(message)
+                            {
+                                RequestId = reqId
+                            };
+                        }
+
+                        if (headers.TryGetValues("x-ratelimit-reset-tokens", out var resetTokensValues))
+                        {
+                            // TPM
+                            var retryAfterAsString = resetTokensValues.FirstOrDefault();
+                            if (TryParseResetTime(retryAfterAsString, out retryAfter) == false)
+                                throw new FormatException($"Unrecognized rate-limit format: '{retryAfterAsString}'");
+                        }
+
+                        if (headers.TryGetValues("x-ratelimit-reset-requests", out var resetRequestsValues))
+                        {
+                            // RPM
+                            var retryAfterAsString = resetRequestsValues.FirstOrDefault();
+                            if (TryParseResetTime(retryAfterAsString, out var retryAfterForReqs) == false)
+                                throw new FormatException($"Unrecognized rate-limit format: '{retryAfterAsString}'");
+
+                            retryAfter = retryAfterForReqs > retryAfter ? retryAfterForReqs : retryAfter;
+                        }
+
+                        // TPM/RPM - should retry only for this exception
+                        throw new GenAiRateLimitException(message)
+                        {
+                            RetryAfter = retryAfter,
+                            RequestId = reqId
+                        };
+                    default:
+                        throw new GenAiTooManyRequestsException(message)
+                        {
+                            RequestId = reqId
+                        };
+                }
+            default:
+                throw new GenUnsuccessfulRequestException(message, response.StatusCode)
+                {
+                    RequestId = reqId
+                };
+        }
+    }
+
+    private string GetRequestId(HttpResponseHeaders headers)
+    {
+        if (headers.TryGetValues("X-Request-ID", out var values) == false || values.IsNullOrEmpty())
+        {
+            return string.Empty;
+        }
+        return  values.FirstOrDefault();
+    }
+
+    private static bool TryParseResetTime(string input, out TimeSpan time)
+    {
+        time = TimeSpan.Zero;
+
+        // As int: 1684293600
+        if (int.TryParse(input, out var seconds1))
+        {
+            time = TimeSpan.FromSeconds(seconds1);
+            return true;
+        }
+
+        // As double: 33011.382867097855
+        if (double.TryParse(input, out var seconds2))
```

Comment:
just to be on the safe side, I would add InvariantCulture

Reply:
Done